tsl/src/nodes/decompress_chunk/planner.c

```diff
@@ -40,169 +41,219 @@ _decompress_chunk_init(void)
+								   chunk_attno - FirstLowInvalidHeapAttributeNumber);
+			}
+			else if (bms_is_member(chunk_attno - FirstLowInvalidHeapAttributeNumber,
+								   chunk_attrs_needed))
+			{
+				destination_attno_in_uncompressed_chunk = chunk_attno;
+				chunk_attrs_found =
+					bms_add_member(chunk_attrs_found,
+								   chunk_attno - FirstLowInvalidHeapAttributeNumber);
+			}
+		}
+		else
+		{
 			/*
-			 * dropped columns have empty string
+			 * Metadata column.
+			 * We always need count column, and sometimes a sequence number
+			 * column. We don't output them, but use them for decompression,
+			 * hence the special negative destination attnos.
+			 * The min/max metadata columns are normally not required for output
+			 * or decompression, they are used only as filter for the compressed
+			 * scan, so we skip them here.
 			 */
-			if (IsA(lfirst(lc), String) && strlen(chunk_col) > 0)
+			Assert(strncmp(column_name,
+						   COMPRESSION_COLUMN_METADATA_PREFIX,
+						   strlen(COMPRESSION_COLUMN_METADATA_PREFIX)) == 0);
+
+			if (strcmp(column_name, COMPRESSION_COLUMN_METADATA_COUNT_NAME) == 0)
 			{
-				tle = make_compressed_scan_targetentry(path, ht_attno, list_length(scan_tlist) + 1);
-				scan_tlist = lappend(scan_tlist, tle);
+				destination_attno_in_uncompressed_chunk = DECOMPRESS_CHUNK_COUNT_ID;
+				missing_count = false;
+			}
+			else if (path->needs_sequence_num &&
+					 strcmp(column_name, COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME) == 0)
+			{
+				destination_attno_in_uncompressed_chunk = DECOMPRESS_CHUNK_SEQUENCE_NUM_ID;
+				missing_sequence = false;
 			}
 		}
+
+		path->decompression_map =
+			lappend_int(path->decompression_map, destination_attno_in_uncompressed_chunk);
 	}
-	else
+
+	/*
+	 * Check that we have found all the needed columns in the scan targetlist.
+	 * We can't conveniently check that we have all columns for all-row vars, so
+	 * skip attno 0 in this check.
+	 */
+	Bitmapset *attrs_not_found = bms_difference(chunk_attrs_needed, chunk_attrs_found);
+	int bit = bms_next_member(attrs_not_found, 0 - FirstLowInvalidHeapAttributeNumber);
+	if (bit >= 0)
 	{
-		/*
-		 * we only need to find unique varattno references here
-		 * multiple references to the same column will be handled by projection
-		 * we need to include junk columns because they might be needed for
-		 * filtering or sorting
-		 */
-		for (bit = bms_next_member(attrs_used, 0 - FirstLowInvalidHeapAttributeNumber); bit > 0;
-			 bit = bms_next_member(attrs_used, bit))
-		{
-			/* bits are offset by FirstLowInvalidHeapAttributeNumber */
-			AttrNumber ht_attno = bit + FirstLowInvalidHeapAttributeNumber;
+		elog(ERROR,
+			 "column '%s' (%d) not found in the scan targetlist for compressed chunk '%s'",
+			 get_attname(path->info->chunk_rte->relid,
+						 bit + FirstLowInvalidHeapAttributeNumber,
+						 /* missing_ok = */ true),
+			 bit + FirstLowInvalidHeapAttributeNumber,
+			 get_rel_name(path->info->compressed_rte->relid));
+	}
```

Comment:
Hmm.... suspect it is not easy to create a test for this, but checking. Can you?

Reply:
This would be a logic error, so hopefully it's not possible to trigger :) The targetlist for the compressed chunk should have all the required columns.

Reply:
Ok.